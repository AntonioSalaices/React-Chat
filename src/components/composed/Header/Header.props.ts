import { Func } from '@Utils/types';

export interface HeaderProps {
  /**
   * Boolean to despite if the user is already logged
   */
  logged: boolean;
  /**
   * Callback that is called when the  user wants to logout
   */
  onClickLogout: Func;
  /**
   * Callback to display the login modal once the user press it
   */
  onClickLoginNavigation: Func;
}
