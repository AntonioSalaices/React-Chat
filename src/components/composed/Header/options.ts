import { LinkType } from '@Utils/types';

export const headerOptions: LinkType[] = [
  { path: '/', name: 'header.links.home', sessionRequired: false },
  { path: 'library', name: 'header.links.library', sessionRequired: false },
  { path: 'chat', name: 'header.links.chat', sessionRequired: true },
  { path: 'dashboard', name: 'header.links.dashboard', sessionRequired: true },
];
