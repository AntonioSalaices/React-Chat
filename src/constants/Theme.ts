export enum Theme {
  Theme = 'theme',
  Light = 'light',
  Dark = 'dark',
}
