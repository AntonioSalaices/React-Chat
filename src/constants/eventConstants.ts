export enum Events {
  LANGUAGE = 'languagechange',
  RESIZE = 'resize',
}

export enum MenuStatus {
  Open = 'open',
  Closed = 'closed',
}
