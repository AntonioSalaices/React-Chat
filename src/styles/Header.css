
.header-container {
    background-color: var(--primary-color);
    height: 60px;
    width: 100%;
    display: flex;
}