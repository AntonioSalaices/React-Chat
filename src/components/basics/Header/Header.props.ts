import { Theme } from "@Constans/Theme";

export interface HeaderProps {
  theme: Theme;
  onChange: (args: Theme) => void;
  icon: JSX.Element;
}
