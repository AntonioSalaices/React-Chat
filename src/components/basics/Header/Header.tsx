import { useEffect, useState } from 'react';
import { Theme } from '@Constans/Theme';
import { translate } from '../../../i18n';

import useThrottle from '@Hooks/useThrottle/useThrottle';

import { Events } from '@Constans/eventConstants';
import { ScreenSize } from '@Constans/screenConstants';
import { HeaderProps } from './Header.props';

import Formatter from '@Utils/formatter';
import { FaLaptop, FaMobileAlt } from 'react-icons/fa';

const { sizeToRange } = Formatter;

const iconsScreen: Record<string, JSX.Element> = {
  [ScreenSize.XS]: <FaMobileAlt />,
  [ScreenSize.SM]: <FaMobileAlt />,
  [ScreenSize.MD]: <FaLaptop />,
  [ScreenSize.LG]: <FaLaptop />,
  [ScreenSize.XL]: <FaLaptop />,
};

const Header: React.FC<HeaderProps> = ({ onChange, theme }) => {
  const selectedTheme: Theme = theme === Theme.DARK ? Theme.LIGHT : Theme.DARK;
  const [range, setRange] = useState<string>(sizeToRange(window.innerWidth));

  useEffect(() => {
    window.addEventListener(Events.RESIZE, handleWindowResize);

    return () => {
      window.removeEventListener(Events.RESIZE, handleWindowResize);
    };
  }, []);

  const handleWindowResize = useThrottle(
    () => {
      setRange(sizeToRange(window.innerWidth));
    },
    100,
    []
  );

  return (
    <nav className="navbar-primary text-white mb-4 justify-between">
      <div className="container">
        <h2 className="site-title">{translate('header.title')}</h2>
        <ul className="display-f justify-center align-center gap-2">
          <li>{iconsScreen[range]}</li>
          <li>
            <button className="btn-light text-dark font-md" onClick={() => onChange(selectedTheme)}>
              {theme}
            </button>
          </li>
        </ul>
      </div>
    </nav>
  );
};
export default Header;
