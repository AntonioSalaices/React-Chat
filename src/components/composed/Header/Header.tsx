import { useEffect, useState } from 'react';
import classNames from 'classnames';

import { translate } from '../../../i18n';

import useThrottle from '@Hooks/useThrottle/useThrottle';
import Formatter from '@Utils/formatter';

import { Events, MenuStatus } from '@Constans/eventConstants';
import { ScreenIcons, AppIcons, AppIcon } from '@Constans/icons';

import { HeaderProps } from './Header.props';
import { headerOptions } from './options';
import { Link } from 'react-router-dom';
import { LinkType } from '@Utils/types';

const { sizeToRange } = Formatter;

const renderHeaderOptions = (options: LinkType[], logged: boolean) => {
  return options
    .filter((option: LinkType) => (logged ? option.sessionRequired : !option.sessionRequired))
    .map(({ path, name, ...rest }: LinkType, index) => (
      <li key={index.toString()}>
        <Link to={path}>{translate(name)}</Link>
      </li>
    ));
};

const Header: React.FC<HeaderProps> = ({ onClickLogout, logged, onClickLoginNavigation }) => {
  const [range, setRange] = useState<string>(sizeToRange(window.innerWidth));
  const [isOpen, setIsOpen] = useState<boolean>(false);

  const handleOpenMenu = () => {
    setIsOpen((isOpened) => !isOpened);
  };

  useEffect(() => {
    window.addEventListener(Events.Resize, handleWindowResize);

    return () => {
      window.removeEventListener(Events.Resize, handleWindowResize);
    };
  }, []);

  const handleWindowResize = useThrottle(
    () => {
      setRange(sizeToRange(window.innerWidth));
    },
    100,
    []
  );

  const containerClasses = classNames('display-f justify-center align-center gap-2 items-list', {
    [MenuStatus.Open]: isOpen,
    [MenuStatus.Closed]: !isOpen,
  });

  const buttonText = logged ? translate('header.links.logOut') : translate('header.links.signIn');

  return (
    <nav className="navbar-primary text-white mb-4 justify-between">
      <div className="container">
        <h2 className="site-title">
          {translate('header.title')}
          <li onClick={handleOpenMenu} className="nav-icon">
            {AppIcons[AppIcon.Menu]}
          </li>
        </h2>
        <ul className={containerClasses}>
          <li>{ScreenIcons[range]}</li>
          {renderHeaderOptions(headerOptions, logged)}
          <li>
            <button
              data-testid="btn-login-navigate"
              className="btn-light text-dark font-md"
              onClick={() => {
                if (logged) {
                  onClickLogout();
                } else {
                  onClickLoginNavigation();
                }
              }}
            >
              {buttonText}
            </button>
          </li>
        </ul>
      </div>
    </nav>
  );
};
export { Header };
