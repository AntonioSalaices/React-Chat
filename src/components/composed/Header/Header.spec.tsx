import { RenderResult, fireEvent, render } from '@testing-library/react';
import Header from './Header';
import { Theme } from '@Constans/Theme';
import { Events } from '@Constans/eventConstants';

describe('<Header />', () => {
  let component: RenderResult;

  const props = {
    onChange: jest.fn(),
    theme: Theme.LIGHT,
  };

  beforeEach(() => {
    component = render(<Header {...props} />);
  });
  test('should render correctly', () => {
    const { container } = component;

    expect(container).toMatchSnapshot();
  });

  test('should render dark theme', () => {
    const newProps = {
      ...props,
      theme: Theme.DARK,
    };
    const HeaderComponent = render(<Header {...newProps} />);

    const { container } = HeaderComponent;

    expect(container).toMatchSnapshot();
  });

  test('should click on theme button and change selected theme', () => {
    const { getByTestId } = component;

    fireEvent.click(getByTestId('btn-change-theme'));

    expect(props.onChange).toHaveBeenCalled();
  });

  test('should call handleResize when resize event has been fired and display mobile icon', () => {
    Object.defineProperty(window, 'innerWidth', {
      writable: true,
      configurable: true,
      value: 150,
    });

    const newProps = {
      ...props,
      theme: Theme.DARK,
    };
    const HeaderComponent = render(<Header {...newProps} />);

    const { getByTestId } = HeaderComponent;

    window.dispatchEvent(new Event(Events.RESIZE));

    expect(getByTestId('mobile')).toBeVisible();
  });
});
