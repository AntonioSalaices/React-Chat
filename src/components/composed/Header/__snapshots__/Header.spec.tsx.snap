// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Header /> should render correctly 1`] = `
<div>
  <nav
    class="navbar-primary text-white mb-4 justify-between"
  >
    <div
      class="container"
    >
      <h2
        class="site-title"
      >
        <reactcomponent
          height="50"
          width="50"
        />
        <li
          class="nav-icon"
        >
          <svg
            fill="currentColor"
            height="1em"
            stroke="currentColor"
            stroke-width="0"
            viewBox="0 0 448 512"
            width="1em"
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              d="M432 416H16a16 16 0 0 0-16 16v32a16 16 0 0 0 16 16h416a16 16 0 0 0 16-16v-32a16 16 0 0 0-16-16zm0-128H16a16 16 0 0 0-16 16v32a16 16 0 0 0 16 16h416a16 16 0 0 0 16-16v-32a16 16 0 0 0-16-16zm0-128H16a16 16 0 0 0-16 16v32a16 16 0 0 0 16 16h416a16 16 0 0 0 16-16v-32a16 16 0 0 0-16-16zm0-128H16A16 16 0 0 0 0 48v32a16 16 0 0 0 16 16h416a16 16 0 0 0 16-16V48a16 16 0 0 0-16-16z"
            />
          </svg>
        </li>
      </h2>
      <ul
        class="display-f justify-center align-center gap-2 items-list closed"
      >
        <li>
          <svg
            data-testid="desktop"
            fill="currentColor"
            height="1em"
            stroke="currentColor"
            stroke-width="0"
            viewBox="0 0 640 512"
            width="1em"
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              d="M624 416H381.54c-.74 19.81-14.71 32-32.74 32H288c-18.69 0-33.02-17.47-32.77-32H16c-8.8 0-16 7.2-16 16v16c0 35.2 28.8 64 64 64h512c35.2 0 64-28.8 64-64v-16c0-8.8-7.2-16-16-16zM576 48c0-26.4-21.6-48-48-48H112C85.6 0 64 21.6 64 48v336h512V48zm-64 272H128V64h384v256z"
            />
          </svg>
        </li>
        <li>
          <a
            href="/"
          >
            [missing "en-US.header.links.home" translation]
          </a>
        </li>
        <li>
          <a
            href="/library"
          >
            [missing "en-US.header.links.library" translation]
          </a>
        </li>
        <li>
          <button
            class="btn-light text-dark font-md"
            data-testid="btn-login-navigate"
          >
            [missing "en-US.header.links.signIn" translation]
          </button>
        </li>
      </ul>
    </div>
  </nav>
</div>
`;

exports[`<Header /> should render dark theme 1`] = `
<div>
  <nav
    class="navbar-primary text-white mb-4 justify-between"
  >
    <div
      class="container"
    >
      <h2
        class="site-title"
      >
        <reactcomponent
          height="50"
          width="50"
        />
        <li
          class="nav-icon"
        >
          <svg
            fill="currentColor"
            height="1em"
            stroke="currentColor"
            stroke-width="0"
            viewBox="0 0 448 512"
            width="1em"
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              d="M432 416H16a16 16 0 0 0-16 16v32a16 16 0 0 0 16 16h416a16 16 0 0 0 16-16v-32a16 16 0 0 0-16-16zm0-128H16a16 16 0 0 0-16 16v32a16 16 0 0 0 16 16h416a16 16 0 0 0 16-16v-32a16 16 0 0 0-16-16zm0-128H16a16 16 0 0 0-16 16v32a16 16 0 0 0 16 16h416a16 16 0 0 0 16-16v-32a16 16 0 0 0-16-16zm0-128H16A16 16 0 0 0 0 48v32a16 16 0 0 0 16 16h416a16 16 0 0 0 16-16V48a16 16 0 0 0-16-16z"
            />
          </svg>
        </li>
      </h2>
      <ul
        class="display-f justify-center align-center gap-2 items-list closed"
      >
        <li>
          <svg
            data-testid="desktop"
            fill="currentColor"
            height="1em"
            stroke="currentColor"
            stroke-width="0"
            viewBox="0 0 640 512"
            width="1em"
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              d="M624 416H381.54c-.74 19.81-14.71 32-32.74 32H288c-18.69 0-33.02-17.47-32.77-32H16c-8.8 0-16 7.2-16 16v16c0 35.2 28.8 64 64 64h512c35.2 0 64-28.8 64-64v-16c0-8.8-7.2-16-16-16zM576 48c0-26.4-21.6-48-48-48H112C85.6 0 64 21.6 64 48v336h512V48zm-64 272H128V64h384v256z"
            />
          </svg>
        </li>
        <li>
          <a
            href="/"
          >
            [missing "en-US.header.links.home" translation]
          </a>
        </li>
        <li>
          <a
            href="/library"
          >
            [missing "en-US.header.links.library" translation]
          </a>
        </li>
        <li>
          <button
            class="btn-light text-dark font-md"
            data-testid="btn-login-navigate"
          >
            [missing "en-US.header.links.signIn" translation]
          </button>
        </li>
      </ul>
    </div>
  </nav>
</div>
`;
